oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/query/MetadataQueryEsDAO.java

```diff
@@ -143,55 +128,55 @@ public Service searchService(final NodeType nodeType, final String serviceCode)
     }
 
     @Override
-    public List<Endpoint> searchEndpoint(String keyword, String serviceId, int limit)
-        throws IOException {
-        final String index = IndexController.LogicIndicesRegister.getPhysicalTableName(
-            EndpointTraffic.INDEX_NAME);
+    public List<ServiceInstance> listInstances(long startTimestamp, long endTimestamp,
+                                               String serviceId) throws IOException {
+        final String index =
+            IndexController.LogicIndicesRegister.getPhysicalTableName(InstanceTraffic.INDEX_NAME);
 
+        final long minuteTimeBucket = TimeBucket.getMinuteTimeBucket(startTimestamp);
         final BoolQueryBuilder query =
             Query.bool()
-                 .must(Query.term(EndpointTraffic.SERVICE_ID, serviceId));
-
-        if (!Strings.isNullOrEmpty(keyword)) {
-            String matchCName = MatchCNameBuilder.INSTANCE.build(EndpointTraffic.NAME);
-            query.must(Query.match(matchCName, keyword));
-        }
-
-        final SearchBuilder search = Search.builder().query(query).size(limit);
+                 .must(Query.range(InstanceTraffic.LAST_PING_TIME_BUCKET).gte(minuteTimeBucket))
+                 .must(Query.term(InstanceTraffic.SERVICE_ID, serviceId));
+        final SearchBuilder search = Search.builder().query(query).size(queryMaxSize);
 
         final SearchResponse response = getClient().search(index, search.build());
-
-        List<Endpoint> endpoints = new ArrayList<>();
-        for (SearchHit searchHit : response.getHits()) {
-            Map<String, Object> sourceAsMap = searchHit.getSource();
-
-            final EndpointTraffic endpointTraffic =
-                new EndpointTraffic.Builder().storage2Entity(sourceAsMap);
-
-            Endpoint endpoint = new Endpoint();
-            endpoint.setId(endpointTraffic.id());
-            endpoint.setName((String) sourceAsMap.get(EndpointTraffic.NAME));
-            endpoints.add(endpoint);
-        }
-
-        return endpoints;
+        return buildInstances(response);
     }
 
     @Override
-    public List<ServiceInstance> getServiceInstances(long startTimestamp, long endTimestamp,
-                                                     String serviceId) throws IOException {
+    public ServiceInstance getInstance(final String instanceId) throws IOException {
         final String index =
             IndexController.LogicIndicesRegister.getPhysicalTableName(InstanceTraffic.INDEX_NAME);
-
-        final long minuteTimeBucket = TimeBucket.getMinuteTimeBucket(startTimestamp);
         final BoolQueryBuilder query =
             Query.bool()
-                 .must(Query.range(InstanceTraffic.LAST_PING_TIME_BUCKET).gte(minuteTimeBucket))
-                 .must(Query.term(InstanceTraffic.SERVICE_ID, serviceId));
-        final SearchBuilder search = Search.builder().query(query).size(queryMaxSize);
+                 .must(Query.term("_id", instanceId));
```

Comment:
How can you find the instance by `_id`? I don't think we ever use `_id` (which is generated by ES) in our logic

Reply:
I think we don't use `ES` generated ID. I haven't verified this method, but AFAIK, `_id` of index is really generally by traffic and metric entities, they have `id()` method for this. Such as

```java
    @Override
    public InsertRequest prepareBatchInsert(Model model, Metrics metrics) {
        Map<String, Object> builder =
            IndexController.INSTANCE.appendMetricTableColumn(model, storageBuilder.entity2Storage(metrics));
        String modelName = TimeSeriesUtils.writeIndexName(model, metrics.getTimeBucket());
        String id = IndexController.INSTANCE.generateDocId(model, metrics.id());
        return getClient().prepareInsert(modelName, id, builder);
    }
```

Reply:
OK make sense to me, we don't use ID generated by ES, we specify them when inserting documents.